text/0000-grid-primitives.md

```diff
@@ -0,0 +1,761 @@
+
+- If a Schema does not exist, the transaction is invalid.
+- If the property list is empty, the transaction is invalid.
+- If one of the new properties has the same name as a property already defined
+  in the scheme, the  transaction is invalid.
+- The signer of the transaction must be an agent in the Pike state and must
+  belong to an organization in Pike state, otherwise the transaction is invalid.
+- The signer of the transaction must belong to the same organization matching
+  the `owner` of the schema, otherwise the transaction is invalid.
+- The agent must have the permission `can_update_schema` for the organization,
+  otherwise the transaction is invalid.
+
+The inputs for SchemaUpdateAction must include:
+
+- Address of the Agent submitting the transaction
+- Address of the Schema  
+
+The outputs for SchemaCreateAction must include:
+
+- Address of the Schema  
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+In this RFC, validating a collection of property values in a data structure
+against a Schema is left to the smart contract implementer. This may result in
+inconsistent validation rules, particularly in the cases where the natural
+default value provided by protobuf deserialization may be ambiguous (for
+example, the enum case).
+
+The dynamic nature of the property values also require a query for the schema in
+order to validate the state.  This incurs an additional cost when creating and
+updating a collection of property values. 
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+This design provides a reusable pattern for arbitrary properties in data
+structures defined in Grid smart contracts. Providing a common language across
+all data structures built with the schemas will allow for the creation of a
+common set of tools, such as REST APIs and Sawtooth state delta export.
+
+Alternatively, the reusable components could be implemented at the message
+level, but this provides little in the way of pre-defined primitives.  They
+would be too broadly defined as protobuf primitive values.  It also moves the
+schema out as an external definition, which would make dynamic applications (for
+example, common UI components) more difficult.  The applications would need to
+inspect the message format itself, in order to provide details about the data
+structures.
+
+# Prior Art
+[prior-art]: #prior-art
+
+These types are adapted from the original [Sawtooth Supply Chain Transaction
+Family
+Specification](https://github.com/hyperledger/sawtooth-supply-chain/blob/master/docs/source/family_specification.rst)
+and the expanded data types RFC, [Sawtooth RFC
+0013](https://github.com/hyperledger/sawtooth-rfcs/blob/master/text/0013-supply-chain-expand-data-types.md).
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+It would be nice to provide validation functions to the Grid SDK that would
+validate that the data structure matches a schema. This would make use the
```

Comment:
The sentence "This would make use the ..." seems to have a couple typos.

Reply:
Fixed